### ISSUE TITLE
Allow adding `version` field to avoid homebrew incorrectly detecting the version

### ISSUE THREAD

**Issue by cdzombak** [Justintime50/homebrew-releaser#39]
I have a formula [here](https://github.com/cdzombak/homebrew-oss/blob/main/formula/runner.rb), managed by homebrew-releaser.

The latest release is 2.8.4, and I have 2.8.3 installed on my machine. However, Homebrew thinks my machine has version "64" installed. I assume it did this because the download URL contains `arm64`.

```
$ runner -version
2.8.3

$ brew upgrade runner
Warning: cdzombak/oss/runner 64 already installed

$ brew info runner
==> cdzombak/oss/runner: stable 64
Lightweight wrapper for better cron jobs and containers
https://github.com/cdzombak/runner
/opt/homebrew/Cellar/runner/64 (3 files, 7MB) *
  Built from source on 2023-11-21 at 12:28:12
From: https://github.com/cdzombak/homebrew-oss/blob/HEAD/Formula/runner.rb
License: LGPL-3.0
```

Perhaps homebrew-releaser could support adding [a `version` field](https://rubydoc.brew.sh/Formula#version-class_method) to the generated formula, so Homebrew is guaranteed to use the correct version rather than incorrectly auto-detecting it?

**Comment by Justintime50** [comment]
How interesting, thanks for reporting this! After some quick poking, I'm not entirely sure why this isn't working as intended. The formula looks like it's generated correctly and by doing `runner --version` locally I get `2.8.4` but when doing `brew info runner` I also get `stable 64`.

Initially I didn't include a version field because with the inclusion of `version` in the formula, it can't pass `brew audit` as Homebrew insists on trying to automatically get the right version of this. I'm open to it but would want to determine if there is a best of both worlds as the tool attempts to pass `brew audit` with how it sets up formula. I'm open to additional insight here. I'll try to look into this in the coming days but can't promise a quick resolution due to the holidays.

**Comment by cdzombak** [comment]
> How interesting, thanks for reporting this! After some quick poking, I'm not entirely sure why this isn't working as intended. The formula looks like it's generated correctly and by doing `runner --version` locally I get `2.8.4` but when doing `brew info runner` I also get `stable 64`.

Agreed: this seems like an odd thing for Homebrew to get wrong, but somehow it is. To be clear, I _don't_ think this is a bug in homebrew-releaser!

> Initially I didn't include a version field because with the inclusion of `version` in the formula, it can't pass `brew audit` as Homebrew insists on trying to automatically get the right version of this. I'm open to it but would want to determine if there is a best of both worlds as the tool attempts to pass `brew audit` with how it sets up formula. I'm open to additional insight here.

I'm definitely a novice with Homebrew formulae, but the formula passes `brew audit --strict --online cdzombak/oss/runner` if I apply the following diff which adds the `version` field:

```diff
diff --git a/formula/runner.rb b/formula/runner.rb
index 1fbb2ab..95611bd 100644
--- a/formula/runner.rb
+++ b/formula/runner.rb
@@ -7,7 +7,8 @@ class Runner < Formula
   homepage "https://github.com/cdzombak/runner"
   url "https://github.com/cdzombak/runner/archive/refs/tags/v2.8.4.tar.gz"
   sha256 "3b016ed3af0f54d5f95f6925264723069b64054be5378d82bdbcc73bf19b910b"
-  license "LGPL-3.0"
+  license "LGPL-3.0-or-later"
+  version "2.8.4"
 
   on_macos do
     on_intel do

```

(Amusingly, the formula as committed in my `homebrew-oss` repo does not pass because Homebrew doesn't like the `license "LGPL-3.0"` value. Again, that's _not_ a homebrew-releaser bug.)

Removing and reinstalling with that updated formula seems to work as expected:

```
$ brew install cdzombak/oss/runner
==> Fetching cdzombak/oss/runner
==> Downloading https://github.com/cdzombak/runner/releases/download/v2.8.4/runner-2.8.4-darwin-arm64.tar.gz
Already downloaded: /Users/cdzombak/Library/Caches/Homebrew/downloads/5f5811b2beacf37285574781e139565670b52a911fa637fba584723a0b3df6d2--runner-2.8.4-darwin-arm64.tar.gz
==> Installing runner from cdzombak/oss
🍺  /opt/homebrew/Cellar/runner/2.8.4: 3 files, 7MB, built in 1 second
==> Running `brew cleanup runner`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

$ brew info runner
==> cdzombak/oss/runner: stable 2.8.4
Lightweight wrapper for better cron jobs and containers
https://github.com/cdzombak/runner
/opt/homebrew/Cellar/runner/2.8.4 (3 files, 7MB) *
  Built from source on 2023-12-01 at 12:30:13
From: https://github.com/cdzombak/homebrew-oss/blob/HEAD/Formula/runner.rb
License: LGPL-3.0-or-later
```

I suspect that `brew audit` checks whether its automatically determined version differs from the formula's declared `version`, and only raises an error if they're the same. Adding an explicit `version` to another of my formulas, for which Homebrew is able to determine the version correctly, raises this audit error:

```
$ brew audit --strict --online cdzombak/oss/listening
cdzombak/oss/listening
  * Stable: version 1.0.3 is redundant with version scanned from URL
Error: 1 problem in 1 formula detected.
```

I think the easiest path forward would be to allow optionally adding the `version` field, but default the option to `false`, and include a note that including it unnecessarily will cause audit failures.

An alternative would be for homebrew-releaser to run Homebrew's "get version from URL" logic and add the `version` field to the formula only when needed. That would avoid the need to add an option, at the cost of significantly increased scope and the effort needed to keep the logic in sync with any future Homebrew changes.

> I'll try to look into this in the coming days but can't promise a quick resolution due to the holidays.

Of course, understood. This is not urgent :)

If you like either of the approaches I outlined above, I'm also happy to implement it myself and send a PR.

**Comment by Justintime50** [comment]
I've put up #40 to address this request, mind checking my work to see if this will do the trick for your workflow? I assume you'd just dynamically set the version on each release workflow? If so, this should do the trick!

**Comment by Justintime50** [comment]
Alright, give `v0.18.0` or `v1` a whirl and use the `version` param and let me know if that doesn't work out for you!

**Comment by cdzombak** [comment]
Thanks, @Justintime50. Testing it out now …

**Comment by cdzombak** [comment]
I just tried to make a new release using the explicit `version` field, and I got this result:

https://github.com/cdzombak/runner/actions/runs/7441908913/job/20244721446

It looks like perhaps the `version` field is assumed to also be the release's tag name? But in my repo, the tag name is `v2.9.0`.

I'll try to dig in and see what's happening and if there's an easy workaround for me,

**Comment by Justintime50** [comment]
That's correct, though I also tag mine with a leading `v`. It appears you didn't pass `v2.9.0` to the version param? It should work if you add the `v` to that field since it won't strip the v till a bit later in the process.

Let me know if that does the trick.

**Comment by cdzombak** [comment]
The good news, it produced an installable formula!

The bad news is that Homebrew is _extremely_ picky and [says](https://github.com/cdzombak/homebrew-oss/actions/runs/7494990859),

```
  /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/cdzombak/homebrew-oss/formula/runner.rb:10:3: C: [Correctable] FormulaAudit/ComponentsOrder: version (line 10) should be put before sha256 (line 9)
```

**Comment by Justintime50** [comment]
Wild! I audit the generated formula on CI and when the `version` param was added this wasn't a rule but apparently is now. I've corrected this and released `v0.18.1` (or v1) with the lint/audit fix! Thanks for sticking with this.

**Comment by cdzombak** [comment]
> Wild! I audit the generated formula on CI and when the `version` param was added this wasn't a rule but apparently is now. I've corrected this and released `v0.18.1` (or v1) with the lint/audit fix! Thanks for sticking with this.

No, thank _you_ for dealing with all this! I appreciate it 😄
